common/junit-platform-native/src/main/java/org/graalvm/junit/platform/JUnitPlatformFeature.java

```diff
@@ -156,20 +157,31 @@ private TestPlan registerTestPlan(Launcher launcher, List<Path> classpath) {
                 .map(Optional::get)
                 .filter(ClassSource.class::isInstance)
                 .map(ClassSource.class::cast)
-                .forEach((ClassSource classSource) -> {
-                    Class<?> clazz = classSource.getJavaClass();
-                    if (debug) {
-                        System.out.println("[Debug] Found test class: " + clazz);
-                    }
-                    RuntimeReflection.registerForReflectiveInstantiation(clazz);
-                    for (Field field : clazz.getDeclaredFields()) {
-                        RuntimeReflection.register(field);
-                    }
-                    for (Method method : clazz.getDeclaredMethods()) {
-                        RuntimeReflection.register(method);
-                    }
-                });
+                .forEach(this::registerTestClassForReflection);
 
         return testPlan;
     }
+
+    private void registerTestClassForReflection(ClassSource classSource) {
+        Class<?> clazz = classSource.getJavaClass();
+        if (debug) {
+            System.out.println("[Debug] Found test class: " + clazz);
+        }
+
+        Class<?> declaringClass = clazz.getEnclosingClass();
+        if (declaringClass == null || Modifier.isStatic(clazz.getModifiers())) {
+            /* A top level or a static nested class must have a no-arg constructor. */
```

Comment:
That's not entirely true.

It is up to the `TestEngine` implementation to determine how the test class is instantiated.

For example, JUnit Jupiter only requires that a test class have a single constructor. Whether or not that single constructor accepts arguments is up to the test author. See [Dependency Injection for Constructors and Methods](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dependency-injection) for details.

Other `TestEngine` implementations may support multiple constructors as they see fit.

Long story, short: we should register all declared constructors for reflection.

Reply:
Ah, that's a good point, it makes sense that JUnit Jupiter would support dependency injection and that other engines could place their own requirements on the test classes. I agree, we should register all declared constructors for reflection.